File: crates/netcfgd-sys/src/watch.rs
//! Watching the config directory, whichever way the system allows.
//!
//! inotify where it works, mtime polling where it does not. The fallback is
//! not defensive programming for its own sake: `inotify_init1` fails with
//! `EMFILE` when `fs.inotify.max_user_instances` is exhausted, which happens
//! on real machines running enough watchers, and some container runtimes and
//! hardened kernels restrict it outright. A config daemon that stops noticing
//! config changes because a limit somewhere else was reached is a worse
//! outcome than one that polls.
//!
//! Both paths answer the same question -- "did anything change?" -- and both
//! answer it the same way netlink does: by saying that something moved, not by
//! saying what. The caller re-reads and recompiles, so a missed detail costs
//! nothing and a missed *event* is what matters.

use crate::inotify::{mask, Inotify};
use std::io;
use std::path::{Path, PathBuf};
use std::time::SystemTime;

/// Which mechanism a watcher ended up with.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Mechanism {
	/// The kernel tells us.
	Inotify,
	/// We ask, repeatedly.
	Polling,
}

impl Mechanism {
	/// A name for logs and for `ncfg status`, because an operator debugging a
	/// reload that did not happen needs to know which one is in play.
	#[must_use]
	pub fn name(self) -> &'static str {
		match self {
			Self::Inotify => "inotify",
			Self::Polling => "mtime polling",
		}
	}
}

/// Watches a set of directories for any change.
pub struct Watcher {
	inotify: Option<Inotify>,
	paths: Vec<PathBuf>,
	fingerprint: Vec<(PathBuf, Option<SystemTime>)>,
	mechanism: Mechanism,
}

impl Watcher {
	/// Watch these directories, preferring inotify.
	///
	/// Never fails: a directory that does not exist yet is still watched by
	/// the polling path, and its appearance counts as a change. That matters
	/// because `conf.d/` is created the first time somebody writes a drop-in.
	#[must_use]
	pub fn new(paths: &[PathBuf]) -> Self {
		let inotify = Inotify::new().ok().filter(|inotify| {
			// A descriptor with no successful watch is worse than none: it
			// would block forever reporting nothing. Insist on at least one.
			paths
				.iter()
				.filter(|path| path.is_dir())
				.filter_map(|path| inotify.watch(path, mask::CONFIG).ok())
				.count() > 0
		});

		let mechanism = if inotify.is_some() {
			Mechanism::Inotify
		} else {
			Mechanism::Polling
		};

		let mut watcher = Self {
			inotify,
			paths: paths.to_vec(),
			fingerprint: Vec::new(),
			mechanism,
		};
		watcher.fingerprint = watcher.take_fingerprint();
		watcher
	}

	/// Watch these directories without inotify.
	///
	/// Public rather than test-only for two reasons: an operator debugging a
	/// reload that is not happening wants to take inotify out of the picture,
	/// and a fallback that only runs when something else has already gone
	/// wrong is a fallback nobody has ever seen work. The tests use it to
	/// exercise both paths against the same assertions.
	#[must_use]
	pub fn polling(paths: &[PathBuf]) -> Self {
		let mut watcher = Self {
			inotify: None,
			paths: paths.to_vec(),
			fingerprint: Vec::new(),
			mechanism: Mechanism::Polling,
		};
		watcher.fingerprint = watcher.take_fingerprint();
		watcher
	}

	/// Which mechanism is in use.
	#[must_use]
	pub fn mechanism(&self) -> Mechanism {
		self.mechanism
	}

	/// Wait up to `timeout_ms` for a change.
	///
	/// Returns false on timeout, so a caller can use this as its own tick.
	///
	/// # Errors
	///
	/// Returns the underlying `io::Error` only where the watch itself failed
	/// in a way that is not recoverable by looking again.
	pub fn wait(&mut self, timeout_ms: i32) -> io::Result<bool> {
		if let Some(inotify) = &self.inotify {
			let events = inotify.wait(timeout_ms)?;
			if events.is_empty() {
				return Ok(false);
			}
			// The fingerprint is kept current even on the inotify path, so
			// that a later fall back to polling does not immediately report a
			// change that was already handled.
			self.fingerprint = self.take_fingerprint();
			return Ok(true);
		}

		std::thread::sleep(std::time::Duration::from_millis(
			u64::try_from(timeout_ms.max(0)).unwrap_or(0),
		));
		let current = self.take_fingerprint();
		if current == self.fingerprint {
			return Ok(false);
		}
		self.fingerprint = current;
		Ok(true)
	}

	/// Every watched path and every `.conf` beneath it, with its mtime.
	///
	/// A file that does not exist contributes `None` rather than being
	/// omitted, so its appearance and its disappearance both change the
	/// fingerprint. Omitting it would make a deleted drop-in invisible.
	fn take_fingerprint(&self) -> Vec<(PathBuf, Option<SystemTime>)> {
		let mut out = Vec::new();
		for path in &self.paths {
			out.push((path.clone(), mtime(path)));
			let Ok(entries) = std::fs::read_dir(path) else {
				continue;
			};
			let mut children: Vec<PathBuf> = entries
				.filter_map(Result::ok)
				.map(|entry| entry.path())
				.collect();
			// Sorted, or the fingerprint would differ run to run on directory
			// order alone and every tick would look like a change.
			children.sort();
			for child in children {
				let stamp = mtime(&child);
				out.push((child, stamp));
			}
		}
		out
	}
}

fn mtime(path: &Path) -> Option<SystemTime> {
	std::fs::metadata(path)
		.ok()
		.and_then(|metadata| metadata.modified().ok())
}

File: c/src/sys/watch.c
/*
 * watch.c -- which mechanism answers "did the configuration change?", and the
 * fingerprint the slow one compares.
 *
 * No syscall that is not a `stat` or a `readdir` is made here: the descriptor
 * and the events are next door in inotify.c, and this file chooses between them
 * and the fallback. watch.h has the argument for there being a fallback at all.
 */
#include "ncfg/watch.h"

#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <time.h>

const char *ncfg_watch_mechanism_name(ncfg_watch_mechanism_t mechanism)
{
	switch (mechanism) {
	case NCFG_WATCH_INOTIFY:
		return "inotify";
	case NCFG_WATCH_POLLING:
		return "mtime polling";
	}
	return "unknown";
}

ncfg_watch_mechanism_t ncfg_watch_mechanism(const ncfg_watch_t *watch)
{
	return watch ? watch->mechanism : NCFG_WATCH_POLLING;
}

static void marks_free(ncfg_watch_marks_t *marks)
{
	size_t at;

	if (!marks) {
		return;
	}
	for (at = 0; at < marks->count; at++) {
		free(marks->items[at].path);
	}
	free(marks->items);
	marks->items = NULL;
	marks->count = 0;
	marks->capacity = 0;
}

/* One path and its mtime, or that it is not there. Absence is a mark rather
 * than a missing entry: see watch.h. */
static int mark_add(ncfg_watch_marks_t *marks, const char *path)
{
	struct stat found;
	char       *copy;

	if (marks->count == marks->capacity) {
		size_t             wanted = marks->capacity ? marks->capacity * 2u : 16u;
		ncfg_watch_mark_t *grown = realloc(marks->items, wanted * sizeof(*grown));

		if (!grown) {
			return 0;
		}
		marks->items = grown;
		marks->capacity = wanted;
	}
	copy = strdup(path);
	if (!copy) {
		return 0;
	}
	memset(&marks->items[marks->count], 0, sizeof(marks->items[0]));
	marks->items[marks->count].path = copy;
	if (stat(path, &found) == 0) {
		marks->items[marks->count].has_mtime = 1;
		marks->items[marks->count].seconds = (int64_t)found.st_mtime;
		marks->items[marks->count].nanoseconds = found.st_mtim.tv_nsec;
	}
	marks->count++;
	return 1;
}

static int compare_marks(const void *left, const void *right)
{
	const ncfg_watch_mark_t *one = (const ncfg_watch_mark_t *)left;
	const ncfg_watch_mark_t *two = (const ncfg_watch_mark_t *)right;

	return strcmp(one->path, two->path);
}

/*
 * Every watched path and everything directly beneath it, with its mtime.
 *
 * **Into a fingerprint of its own**, never into the watcher's: the one being
 * compared against has to survive being compared, and a version of this that
 * cleared the watcher first reported a change on every tick -- which is how it
 * was found.
 *
 * The children are sorted, or the fingerprint would differ run to run on
 * directory order alone -- which is the filesystem's and is not stable -- and
 * every tick would look like a change for a second reason.
 *
 * Everything beneath, not only what looks like configuration: a drop-in renamed
 * from `netcfgd.conf.new` into place is a change, and a fingerprint that only
 * counted the final name would see the rename and nothing before it. The Rust's
 * doc comment says `.conf` and its code does the same as this; the code is what
 * the behaviour was measured against.
 */
static int fingerprint(const ncfg_watch_t *watch, ncfg_watch_marks_t *out)
{
	size_t at;

	memset(out, 0, sizeof(*out));
	for (at = 0; at < watch->path_count; at++) {
		DIR           *directory;
		struct dirent *entry;
		char           child[4096];
		size_t         first;

		if (!mark_add(out, watch->paths[at])) {
			marks_free(out);
			return 0;
		}
		directory = opendir(watch->paths[at]);
		if (!directory) {
			/* A directory that is not there yet is a mark of its
			 * own, above, and its appearance is a change. */
			continue;
		}
		first = out->count;
		while ((entry = readdir(directory)) != NULL) {
			int written;

			if (strcmp(entry->d_name, ".") == 0 ||
			    strcmp(entry->d_name, "..") == 0) {
				continue;
			}
			written = snprintf(child, sizeof(child), "%s/%s", watch->paths[at],
			    entry->d_name);
			if (written < 0 || (size_t)written >= sizeof(child)) {
				continue;
			}
			if (!mark_add(out, child)) {
				(void)closedir(directory);
				marks_free(out);
				return 0;
			}
		}
		(void)closedir(directory);
		if (out->count - first > 1u) {
			qsort(&out->items[first], out->count - first,
			    sizeof(out->items[0]), compare_marks);
		}
	}
	return 1;
}

static int same_fingerprint(const ncfg_watch_marks_t *one, const ncfg_watch_marks_t *two)
{
	size_t at;

	if (one->count != two->count) {
		return 0;
	}
	for (at = 0; at < one->count; at++) {
		if (strcmp(one->items[at].path, two->items[at].path) != 0) {
			return 0;
		}
		if (one->items[at].has_mtime != two->items[at].has_mtime) {
			return 0;
		}
		/* Nanoseconds as well as seconds: two writes in one second is
		 * an editor saving twice, and a watcher that could not tell them
		 * apart would miss the second. */
		if (one->items[at].has_mtime &&
		    (one->items[at].seconds != two->items[at].seconds ||
		    one->items[at].nanoseconds != two->items[at].nanoseconds)) {
			return 0;
		}
	}
	return 1;
}

static int paths_copy(ncfg_watch_t *watch, const char *const *paths, size_t count,
    char *err, size_t err_size)
{
	size_t at;

	if (count == 0) {
		return 1;
	}
	watch->paths = calloc(count, sizeof(*watch->paths));
	if (!watch->paths) {
		ncfg_error_set(err, err_size, "out of memory watching for changes");
		return 0;
	}
	for (at = 0; at < count; at++) {
		if (!paths[at]) {
			ncfg_error_set(err, err_size, "a directory to watch cannot be nothing");
			return 0;
		}
		watch->paths[at] = strdup(paths[at]);
		if (!watch->paths[at]) {
			ncfg_error_set(err, err_size, "out of memory watching for changes");
			return 0;
		}
		watch->path_count++;
	}
	return 1;
}

/*
 * Replace the fingerprint with a fresh one, and say nothing about whether it
 * differed.
 *
 * Used both for the first one -- which is what every later one is compared
 * against -- and for the inotify path, which keeps it current without ever
 * comparing it.
 */
static int refresh(ncfg_watch_t *watch, char *err, size_t err_size)
{
	ncfg_watch_marks_t taken;

	if (!fingerprint(watch, &taken)) {
		ncfg_error_set(err, err_size, "out of memory watching for changes");
		return 0;
	}
	marks_free(&watch->marks);
	watch->marks = taken;
	return 1;
}

int ncfg_watch_open_polling(ncfg_watch_t *watch, const char *const *paths, size_t count,
    char *err, size_t err_size)
{
	if (!watch || (count > 0 && !paths)) {
		ncfg_error_set(err, err_size, "watching needs somewhere to keep the watch");
		return 0;
	}
	memset(watch, 0, sizeof(*watch));
	ncfg_inotify_init(&watch->inotify);
	watch->mechanism = NCFG_WATCH_POLLING;
	if (!paths_copy(watch, paths, count, err, err_size)) {
		ncfg_watch_close(watch);
		return 0;
	}
	if (!refresh(watch, err, err_size)) {
		ncfg_watch_close(watch);
		return 0;
	}
	return 1;
}

int ncfg_watch_open(ncfg_watch_t *watch, const char *const *paths, size_t count,
    char *err, size_t err_size)
{
	size_t at;
	size_t watched = 0;

	if (!ncfg_watch_open_polling(watch, paths, count, err, err_size)) {
		return 0;
	}
	/* The descriptor is asked for after the paths are in place, so that a
	 * kernel that refuses one leaves a watcher that is already usable. */
	if (!ncfg_inotify_open(&watch->inotify, NULL, 0)) {
		return 1;
	}
	for (at = 0; at < watch->path_count; at++) {
		if (ncfg_inotify_watch(&watch->inotify, watch->paths[at],
		    NCFG_WATCH_CONFIG_MASK, NULL, NULL, 0)) {
			watched++;
		}
	}
	if (watched == 0) {
		/* A descriptor with no successful watch is worse than none: it
		 * would block for ever reporting nothing. */
		ncfg_inotify_close(&watch->inotify);
		return 1;
	}
	watch->mechanism = NCFG_WATCH_INOTIFY;
	return 1;
}

void ncfg_watch_close(ncfg_watch_t *watch)
{
	size_t at;

	if (!watch) {
		return;
	}
	ncfg_inotify_close(&watch->inotify);
	for (at = 0; at < watch->path_count; at++) {
		free(watch->paths[at]);
	}
	free(watch->paths);
	watch->paths = NULL;
	watch->path_count = 0;
	marks_free(&watch->marks);
	watch->mechanism = NCFG_WATCH_POLLING;
}

int ncfg_watch_wait(ncfg_watch_t *watch, int timeout_ms, int *changed, char *err,
    size_t err_size)
{
	ncfg_watch_marks_t taken;

	if (changed) {
		*changed = 0;
	}
	if (!watch || !changed) {
		ncfg_error_set(err, err_size, "waiting for a change needs somewhere to say so");
		return 0;
	}
	if (watch->mechanism == NCFG_WATCH_INOTIFY) {
		ncfg_inotify_batch_t batch;

		if (!ncfg_inotify_wait(&watch->inotify, timeout_ms, &batch, err, err_size)) {
			return 0;
		}
		if (batch.length == 0) {
			return 1;
		}
		/* The events are not read apart: like a netlink multicast
		 * message, one means "something moved, look again". The caller
		 * re-reads and recompiles, so a missed detail costs nothing and
		 * a missed event is what matters.
		 *
		 * The fingerprint is kept current here so that a later fall back
		 * to polling does not immediately report a change that was
		 * already handled. */
		if (!refresh(watch, err, err_size)) {
			return 0;
		}
		*changed = 1;
		return 1;
	}

	{
		struct timespec wanted;
		long            milliseconds = timeout_ms > 0 ? (long)timeout_ms : 0;

		wanted.tv_sec = milliseconds / 1000L;
		wanted.tv_nsec = (milliseconds % 1000L) * 1000000L;
		(void)nanosleep(&wanted, NULL);
	}
	if (!fingerprint(watch, &taken)) {
		ncfg_error_set(err, err_size, "out of memory watching for changes");
		return 0;
	}
	if (same_fingerprint(&watch->marks, &taken)) {
		marks_free(&taken);
		return 1;
	}
	marks_free(&watch->marks);
	watch->marks = taken;
	*changed = 1;
	return 1;
}
